lib/at_cmd/at_cmd.c

```diff
@@ -111,23 +134,114 @@ static int get_return_code(char *buf, struct return_state_object *ret)
 
 static void callback_worker(struct k_work *item)
 {
-	struct callback_work_item *data =
-		CONTAINER_OF(item, struct callback_work_item, work);
+	ARG_UNUSED(item);
+	struct resp_item resp;
 
-	if (data != NULL) {
-		data->callback(data->data);
+	while (k_msgq_get(&responses, &resp, K_NO_WAIT) == 0) {
+		if (resp.callback != NULL ||
+		    resp.buf != NULL) {
+			resp.callback(resp.buf);
+		}
+		if (resp.flags & AT_RESP_ALLOC) {
+			k_free(resp.buf);
+		}
 	}
+}
+
+
 
-	k_mem_slab_free(&rsp_work_items, (void **)&data);
+/* Distributes an AT response to the appropriate queue */
+static void distribute_resp(const struct cmd_item *cmd, struct resp_item *resp)
+{
+	if (resp->state != AT_CMD_NOTIFICATION &&
+	    cmd->flags & AT_CMD_SYNC) {
+		LOG_DBG("Enqueueing response for sync call");
+		k_msgq_put(&response_sync, resp, K_FOREVER);
+	} else {
+		LOG_DBG("Enqueueing response for callback queue");
+		k_msgq_put(&responses, resp, K_FOREVER);
+		k_work_submit(&callback_work);
+	}
 }
 
+static inline int at_write(const char *const cmd)
+{
+	int bytes_sent;
+	int bytes_to_send = strlen(cmd);
+
+	LOG_DBG("Sending command %s", log_strdup(cmd));
+
+	bytes_sent = send(common_socket_fd, cmd, bytes_to_send, 0);
```

Comment:
Sending a null-terminated String should work.

Reply:
I am not sure I follow this one. Could you elaborate on what is suggested?

Reply:
Handled offline